Introduce subspace-test-runtime and subspace-test-service

liuchengxu opened the pull request:
`subspace-test-runtime` is derived from the current `subspace-runtime` with some different parameters(~~`MILLISECS_PER_BLOCK`, `SLOT_DURATION`~~, `INITIAL_SOLUTION_RANGE`), which is capable of writing more realistic tests (#48 can be fixed) and catching regressions like #265. The test runtime and service will also be used in the cirrus tests.

This PR does not bring in any tests, only the test framework, a sample test was added in `subspace-test-service`.

liuchengxu committed:
Add subspace-test-service

nazar-pc commented on test/subspace-test-runtime/src/lib.rs:
```diff
@@ -119,11 +119,11 @@ pub const SSC: Balance = (10 * SHANNON).pow(DECIMAL_PLACES as u32);
 // NOTE: Currently it is not possible to change the slot duration after the chain has started.
 //       Attempting to do so will brick block production.
-const SLOT_DURATION: u64 = 1000;
+const SLOT_DURATION: u64 = 2000;
```
These will make tests very slow.

liuchengxu replied:
The block production would fail with the normal setting because it would take too long, and even the block time has been doubled, you still need to run the tests in release mode, it still failed if you run it in the debug mode.

nazar-pc replied:
Why would it take too long?

liuchengxu replied:
Because we're still using the slot based mechanism and the machine is not capable of running multiple nodes simultaneously and still being able to produce blocks in a pretty limited time. https://github.com/paritytech/substrate/blob/4cd2cc958f5a078d87e4eae877167c40cf2ebf9e/client/consensus/slots/src/lib.rs#L318

nazar-pc replied:
Why not capable? Node takes a tiny fraction of CPU core to run, you should be able to run 10 nodes on a single core, especially when blockchain starts. Even in debug mode it should be feasible to have many nodes running concurrently on the same core. If not we're doing something wrong. We can't run tests that require many blocks with 0.5 blocks per second speed, it will take forever in CI and even locally. In release mode I was running node with 10 blocks/s speed without issues.

liuchengxu replied:
Well, the argument makes sense, but I don't see anything obviously wrong. I can dig into that if you want, just not sure how much time it will take.

nazar-pc replied:
Please do dig into it a little bit, it shouldn't take anywhere close to a second to produce a block.

liuchengxu replied:
Turned out that `runtime_cache_size` was misconfigured, fixed in https://github.com/subspace/subspace/pull/266/commits/89795e18cd05993d3f38b23a160bc54cb1f401e9

nazar-pc reviewed (commented):
A few observations right away. If it is derived from regular crates, I would prefer to see first commit that copies files unmodified and second that changes it to the desired shape, that way it is way easier to review. This is both about runtime and service. As to chainspec and farming, I think it needs to be in a separate crate, probably `subspace-test-client`?

liuchengxu committed:
Extract `subspace-test-client` from `subspace-test-service`

liuchengxu committed:
Increase `runtime_cache_size` in node config

Otherwise each runtime api will construct a new runtime instance and
that take 500ms on my machine, making the block production too slow.

Fixes https://github.com/subspace/subspace/pull/266#discussion_r810601337

nazar-pc committed:
Copy `crates/subspace-runtime` into `test/subspace-test-runtime` with zero changes

nazar-pc committed:
Add subspace-test-runtime

nazar-pc committed:
Remove unnecessary test runtime dependencies

nazar-pc reviewed (approved):
Pushed one commit in the front of everything so it is very clear what changes were done to the runtime and removed benchmark dependencies in the last commit as unused.

Why not having these at `crates` though?
